### PR TITLE
Support `pre` execution for action

### PR DESCRIPTION
**Pull request by TingluoHuang**
https://github.com/actions/runner/pull/361

### CODE REVIEW

File: src/Runner.Worker/ActionManifestManager.cs [hunk 1]
```diff
@@ -340,6 +352,8 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
                             Arguments = argsToken,
                             EntryPoint = entrypointToken?.Value,
                             Environment = envToken,
+                            Init = preEntrypointToken?.Value,
+                            InitCondition = preIfToken?.Value ?? "always()",
```
Comment (TingluoHuang) [inline src/Runner.Worker/ActionManifestManager.cs]: needs agree on what's the default condition.

File: src/Runner.Worker/ActionManifestManager.cs [hunk 2]
```diff
@@ -356,6 +370,8 @@ public ActionDefinitionData Load(IExecutionContext executionContext, string mani
                         return new NodeJSActionExecutionData()
                         {
                             Script = mainToken.Value,
+                            Init = preToken?.Value,
+                            InitCondition = preIfToken?.Value ?? "always()",
```
Comment (TingluoHuang) [inline src/Runner.Worker/ActionManifestManager.cs]: needs agree on what's the default condition.

File: src/Runner.Worker/ActionManifestManager.cs (hunk 1 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ActionManifestManager.cs]: stick with `always()` for now base on ADR.

File: src/Runner.Worker/ActionManifestManager.cs (hunk 2 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ActionManifestManager.cs]: stick with `always()` for now base on ADR.

**Comment by ericsciple** [comment]
Should we have another engineer from our team approve the ADR? Hesitant to approve the code into master otherwise. I'm personally not a fan of the feature - but comfortable if out-voted :)

File: src/Runner.Worker/ActionManager.cs [hunk 3]
```diff
@@ -117,6 +131,22 @@ public async Task<List<JobExtensionRunner>> PrepareActionsAsync(IExecutionContex
                             imagesToBuildInfo[setupInfo.ActionRepository] = setupInfo;
                         }
                     }
+
+                    var repoAction = action.Reference as Pipelines.RepositoryPathReference;
+                    if (repoAction.RepositoryType != Pipelines.PipelineConstants.SelfAlias)
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManager.cs:136]: when RepositoryType == SelfAlias does that mean it's a local action (i.e. `uses: ./[...]`)?

File: src/Runner.Worker/ActionManager.cs [hunk 4]
```diff
@@ -117,6 +131,22 @@ public async Task<List<JobExtensionRunner>> PrepareActionsAsync(IExecutionContex
...
+                    {
+                        var definition = LoadAction(executionContext, action);
+                        if (definition.Data.Execution.HasInit)
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManager.cs:139]: For container job, does pre run inside the container?

File: src/Runner.Worker/ActionManager.cs [hunk 5]
```diff
@@ -21,11 +21,24 @@
...
+    {
+        public PrepareResult(List<JobExtensionRunner> containerSetupSteps, Dictionary<Guid, IActionRunner> preStepTracker)
+        {
+            this.ContainerSetupSteps = containerSetupSteps;
+            this.PreStepTracker = preStepTracker;
+        }
+
+        public List<JobExtensionRunner> ContainerSetupSteps { get; set; }
+
+        public Dictionary<Guid, IActionRunner> PreStepTracker { get; set; }
```
Comment (ericsciple) [inline src/Runner.Worker/ActionManager.cs:34]: nit: it looks like different layers uses different terminology (`pre` vs `prepare` vs `init`). Same for `post` vs `cleanup`

File: src/Runner.Worker/ActionRunner.cs [hunk 6]
```diff
@@ -81,20 +82,18 @@ public async Task RunAsync()
...
+                Action.Reference is Pipelines.RepositoryPathReference repoAction &&
+                string.Equals(repoAction.RepositoryType, Pipelines.PipelineConstants.SelfAlias, StringComparison.OrdinalIgnoreCase))
+            {
+                ExecutionContext.Warning($"`pre` execution is not supported for local action from '{repoAction.Path}'");
+            }
+
             // The action has post cleanup defined.
             // we need to create timeline record for them and add them to the step list that StepRunner is using
-            if (handlerData.HasCleanup && Stage == ActionRunStage.Main)
+            if (handlerData.HasCleanup && (Stage == ActionRunStage.Pre || Stage == ActionRunStage.Main))
```
Comment (ericsciple) [inline src/Runner.Worker/ActionRunner.cs:94]: what if the action has pre, main, and post?

Comment (ericsciple) [reply src/Runner.Worker/ActionRunner.cs:94]: will this create two `post` steps?

File: src/Runner.Worker/ExecutionContext.cs [hunk 7]
```diff
@@ -161,6 +161,9 @@ public sealed class ExecutionContext : RunnerService, IExecutionContext
         // Only job level ExecutionContext has PostJobSteps
         public Stack<IStep> PostJobSteps { get; private set; }
 
+        // Only job level ExecutionContext has StepsWithPostRegisted
```
Comment (ericsciple) [inline src/Runner.Worker/ExecutionContext.cs:164]: typo: `StepsWithPostRegisted` =\> `StepsWithPostRegistered`

File: src/Runner.Worker/ExecutionContext.cs [hunk 8]
```diff
@@ -248,9 +251,15 @@ public void ForceTaskComplete()
             });
         }
 
-        public void RegisterPostJobStep(string refName, IStep step)
+        public void RegisterPostJobStep(IStep step)
         {
-            step.ExecutionContext = Root.CreatePostChild(step.DisplayName, refName, IntraActionState);
+            if (step is IActionRunner actionRunner && !Root.StepsWithPostRegisted.Add(actionRunner.Action.Id))
```
Comment (ericsciple) [inline src/Runner.Worker/ExecutionContext.cs:256]: is `Id` the ref name?

Comment (ericsciple) [reply src/Runner.Worker/ExecutionContext.cs:256]: nm, it must be the guid

File: src/Runner.Worker/ActionRunner.cs (hunk 6 above)
Comment (ericsciple) [reply src/Runner.Worker/ActionRunner.cs:94]: nm it looks like RegisterPostJobStep handles gracefully.

File: src/Runner.Worker/ExecutionContext.cs [hunk 9]
```diff
@@ -856,7 +868,8 @@ private IExecutionContext CreatePostChild(string displayName, string refName, Di
                 _childTimelineRecordOrder = _childTimelineRecordOrder * 2;
             }
 
-            return CreateChild(Guid.NewGuid(), displayName, refName, null, null, intraActionState, _childTimelineRecordOrder - Root.PostJobSteps.Count);
+            var newGuid = Guid.NewGuid();
+            return CreateChild(newGuid, displayName, newGuid.ToString("N"), null, null, intraActionState, _childTimelineRecordOrder - Root.PostJobSteps.Count);
```
Comment (ericsciple) [inline src/Runner.Worker/ExecutionContext.cs:872]: does this fix a bug?

File: src/Runner.Worker/JobExtension.cs [hunk 10]
```diff
@@ -260,7 +280,14 @@ public async Task<List<IStep>> InitializeJob(IExecutionContext jobContext, Pipel
                         if (step is IActionRunner actionStep)
                         {
                             ArgUtil.NotNull(actionStep, step.DisplayName);
-                            actionStep.ExecutionContext = jobContext.CreateChild(actionStep.Action.Id, actionStep.DisplayName, actionStep.Action.Name, actionStep.Action.ScopeName, actionStep.Action.ContextName);
+                            if (intraActionStates.ContainsKey(actionStep.Action.Id))
```
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs:283]: trygetvalue might simplify

Comment (ericsciple) [reply src/Runner.Worker/JobExtension.cs:283]: (would be able to eliminate `if`)

File: src/Runner.Worker/JobExtension.cs [hunk 11]
```diff
@@ -239,9 +239,23 @@ public async Task<List<IStep>> InitializeJob(IExecutionContext jobContext, Pipel
...
+                                preStep.DisplayName = $"Pre {preStep.DisplayName}";
+                                preJobSteps.Add(preStep);
+                            }
                         }
                     }
 
+                    var intraActionStates = new Dictionary<Guid, Dictionary<string, string>>();
+                    foreach (var preStep in prepareResult.PreStepTracker)
+                    {
+                        intraActionStates[preStep.Key] = new Dictionary<string, string>();
```
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs:256]: ordinal ignore case

File: src/Runner.Worker/JobExtension.cs [hunk 12]
```diff
@@ -239,9 +239,23 @@ public async Task<List<IStep>> InitializeJob(IExecutionContext jobContext, Pipel
 
                             actionRunner.TryEvaluateDisplayName(contextData, context);
                             jobSteps.Add(actionRunner);
+
+                            if (prepareResult.PreStepTracker.TryGetValue(step.Id, out var preStep))
```
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs:243]: nit: would it be more straight forward for PrepareResult.PreStepTracker to be a list? Then could just simply add to preJobSteps? I find myself jumping back and forth in this file trying to figure out what the dictionary lookup solves.

File: src/Runner.Worker/JobExtension.cs [hunk 13]
```diff
@@ -239,9 +239,23 @@ public async Task<List<IStep>> InitializeJob(IExecutionContext jobContext, Pipel
 
                             actionRunner.TryEvaluateDisplayName(contextData, context);
                             jobSteps.Add(actionRunner);
+
```
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs:242]: It looks like there is an assumption that main is always defined. Is there validation somewhere that enforces?

File: src/Runner.Worker/ActionManager.cs (hunk 5 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ActionManager.cs:34]: i will change everything to be init and cleanup

File: src/Runner.Worker/ActionManager.cs (hunk 3 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ActionManager.cs:136]: yes

File: src/Runner.Worker/ActionManager.cs (hunk 4 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ActionManager.cs:139]: yes, it will

File: src/Runner.Worker/ExecutionContext.cs (hunk 7 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ExecutionContext.cs:164]: fixed.

File: src/Runner.Worker/ExecutionContext.cs (hunk 8 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ExecutionContext.cs:256]: yes. it's the GUID

File: src/Runner.Worker/ExecutionContext.cs (hunk 9 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ExecutionContext.cs:872]: i just stop using `refName`

File: src/Runner.Worker/JobExtension.cs (hunk 12 above)
Comment (TingluoHuang) [reply src/Runner.Worker/JobExtension.cs:243]: I need to tracking which action.id has PreStep, so the `pre` and `main` can share same `intraActionState`

File: src/Runner.Worker/JobExtension.cs (hunk 13 above)
Comment (TingluoHuang) [reply src/Runner.Worker/JobExtension.cs:242]: i have the validation when i parse the action.yml file, it requires the node action to have the `main` and container action to have the `image`

File: src/Runner.Worker/JobExtension.cs (hunk 11 above)
Comment (TingluoHuang) [reply src/Runner.Worker/JobExtension.cs:256]: fixed.

File: src/Runner.Worker/JobExtension.cs (hunk 10 above)
Comment (TingluoHuang) [reply src/Runner.Worker/JobExtension.cs:283]: fixed.

File: src/Runner.Worker/ActionManager.cs (hunk 5 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ActionManager.cs:34]: sorry, `pre` and `post` to lines up with yaml

**Comment by TingluoHuang** [comment]
@juliobbv for checking this PR and [ADR](https://github.com/actions/runner/pull/361)

File: src/Runner.Worker/ExecutionContext.cs [hunk 14]
```diff
@@ -248,9 +251,15 @@ public void ForceTaskComplete()
...
 
-        public void RegisterPostJobStep(string refName, IStep step)
+        public void RegisterPostJobStep(IStep step)
         {
-            step.ExecutionContext = Root.CreatePostChild(step.DisplayName, refName, IntraActionState);
+            if (step is IActionRunner actionRunner && !Root.StepsWithPostRegistered.Add(actionRunner.Action.Id))
+            {
+                Trace.Info($"'post' of '{actionRunner.DisplayName}' already push to post step stack.");
+                return;
+            }
```
Comment (juliobbv) [inline src/Runner.Worker/ExecutionContext.cs:260]: Why do we need to do the check? Do we need to worry about duplicate post steps?

File: src/Test/L0/Listener/MessageListenerL0.cs [hunk 15]
```diff
@@ -660,7 +660,7 @@ public async void CreateSessionWithOriginalGetMessageMigtateToMigrated()
                         _settings.AgentId))
                     .Returns(async () =>
                     {
-                        await Task.Delay(10);
+                        await Task.Delay(100);
```
Comment (juliobbv) [inline src/Test/L0/Listener/MessageListenerL0.cs:663]: What does this fix?

File: src/Test/L0/Worker/ActionManagerL0.cs [hunk 16]
```diff
@@ -682,6 +682,104 @@ public async void PrepareActions_RepositoryActionWithActionfile_Node()
...
+                //Act
+                try
+                {
+                    await _actionManager.PrepareActionsAsync(_ec.Object, actions);
+                }
+                catch (ArgumentException)
+                {
+                    var traceFile = Path.GetTempFileName();
+                    File.Copy(_hc.TraceFileName, traceFile, true);
+                    Assert.Contains("Entry javascript fils is not provided.", File.ReadAllText(traceFile));
```
Comment (juliobbv) [inline src/Test/L0/Worker/ActionManagerL0.cs:719]: `fils` -> `file`

File: src/Test/L0/Worker/ActionManifestManagerL0.cs [hunk 17]
```diff
@@ -323,6 +415,94 @@ public void Load_NodeAction()
             }
         }
 
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
```
Comment (juliobbv) [inline src/Test/L0/Worker/ActionManifestManagerL0.cs:420]: Nice set of new tests! 👍

File: src/Runner.Worker/ExecutionContext.cs (hunk 14 above)
Comment (TingluoHuang) [reply src/Runner.Worker/ExecutionContext.cs:260]: Yes.

**Pull request merged**